ibis/expr/analysis.py

```diff
@@ -198,27 +155,37 @@ def finder(expr):
 
 
 def substitute(fn, expr):
-    """
-    Substitute expressions with other expressions.
-    """
+    """Substitute expressions with other expressions."""
     node = expr.op()
 
     result = fn(node)
     if result is lin.halt:
         return expr
     elif result is not lin.proceed:
-        return result.to_expr()
+        assert isinstance(result, ir.Expr), type(result)
+        return result
 
     new_args = []
     for arg in node.args:
         if isinstance(arg, tuple):
-            arg = tuple(substitute(fn, expr) for expr in arg)
+            arg = tuple(
+                substitute(fn, expr) if isinstance(arg, ir.Expr) else expr
```

Comment:
If `arg` is a tuple then it won't apply the substitution on those values.

These checks must be moved to the top of the substitute() function.